lib/vrl/compiler/src/expression/op.rs

```diff
@@ -93,28 +93,37 @@ impl Expression for Op {
         use ast::Opcode::*;
         use Value::*;
 
-        let lhs = self.lhs.resolve(ctx);
-        let mut rhs = || self.rhs.resolve(ctx);
+        if let Err = self.opcode {
+            return self.lhs.resolve(ctx).or_else(|_| self.rhs.resolve(ctx));
+        } else if let Or = self.opcode {
+            return self
+                .lhs
+                .resolve(ctx)?
+                .try_or(|| self.rhs.resolve(ctx))
+                .map_err(Into::into);
+        };
+
+        let lhs = self.lhs.resolve(ctx)?;
```

Comment:
Why not move this to the top? `lhs` is _always_ resolved.

Reply:
It's always resolved, but for `Err` it's not short-circuited (note the `?`).